src/rust/catnap/win/overlapped.rs

```diff
@@ -0,0 +1,774 @@
+        F2: FnOnce(*mut OVERLAPPED) -> Result<(), Fail>,
+        F3: FnOnce(OverlappedResult) -> Result<R, Fail>,
+    {
+        self.do_io_with(
+            (),
+            yielder,
+            |_, overlapped: *mut OVERLAPPED| -> Result<(), Fail> { start(overlapped) },
+            |_, overlapped: *mut OVERLAPPED| -> Result<(), Fail> { cancel(overlapped) },
+            |_, result: OverlappedResult| -> Result<R, Fail> { finish(result) },
+        )
+        .await
+    }
+
+    /// Process a single overlapped entry.
+    #[allow(unused)]
+    fn process_overlapped(&mut self, entry: &OVERLAPPED_ENTRY) {
+        if let Some(overlapped) = std::ptr::NonNull::new(entry.lpOverlapped) {
+            // Safety: this is valid as long as the caller follows the contract: all queued OVERLAPPED instances are
+            // generated by `IoCompletionPort` API.
+            let overlapped: Pin<&mut OverlappedCompletion> = OverlappedCompletion::unmarshal(overlapped);
+
+            // Safety: the OVERLAPPED does not need to be pinned after being dequeued from the completion port.
+            let overlapped: &mut OverlappedCompletion = unsafe { overlapped.get_unchecked_mut() };
+            if let Some(mut yielder_handle) = overlapped.yielder_handle.take() {
+                debug_assert!(entry.dwNumberOfBytesTransferred as usize == overlapped.overlapped.InternalHigh);
+                overlapped.completion_key = entry.lpCompletionKey;
+                yielder_handle.wake_with(Ok(()));
+            } else {
+                // This can happen due to a failed cancellation or any other error on the do_overlapped path.
+                trace!("I/O dropped");
+                let free_fn: unsafe fn(*mut OVERLAPPED) = overlapped.free;
+                unsafe { (free_fn)(entry.lpOverlapped) };
+            }
+        }
+    }
+
+    /// Process entries by peeking the completion port.
+    #[allow(unused)]
+    pub fn process_events(&mut self) -> Result<(), Fail> {
+        const BATCH_SIZE: usize = 4;
+        let mut entries: [OVERLAPPED_ENTRY; BATCH_SIZE] = [OVERLAPPED_ENTRY::default(); BATCH_SIZE];
+
+        loop {
+            let mut dequeued: u32 = 0;
+            match unsafe { GetQueuedCompletionStatusEx(self.iocp, entries.as_mut_slice(), &mut dequeued, 0, FALSE) } {
+                Ok(()) => {
+                    for i in 0..dequeued {
+                        self.process_overlapped(&entries[i as usize]);
+                    }
+
+                    if dequeued < BATCH_SIZE as u32 {
+                        return Ok(());
+                    }
+                },
+
+                Err(err) if err.code() == WIN32_ERROR(WAIT_TIMEOUT.0).into() => return Ok(()),
+
+                Err(err)
+                    if err.code() == WIN32_ERROR(WAIT_ABANDONED.0).into()
+                        || err.code() == ERROR_INVALID_HANDLE.into() =>
+                {
+                    return Err(Fail::new(libc::EBADF, "completion port closed"))
```

Comment:
Possibly returning the error code might enable the caller/logs to reflect richer context.